x/stake/keeper/params.go

```diff
@@ -0,0 +1,109 @@
+const (
+	DefaultParamSpace = "Stake"
+)
+
+// nolint - Key generators for parameter access
+func KeyInflationRateChange() params.Key { return params.NewKey("InflationRateChange") }
+func KeyInflationMax() params.Key        { return params.NewKey("InflationMax") }
+func KeyInflationMin() params.Key        { return params.NewKey("InflationMin") }
+func KeyGoalBonded() params.Key          { return params.NewKey("GoalBonded") }
+func KeyUnbondingTime() params.Key       { return params.NewKey("UnbondingTime") }
+func KeyMaxValidators() params.Key       { return params.NewKey("MaxValidators") }
+func KeyBondDenom() params.Key           { return params.NewKey("BondDenom") }
+
+// Cached parameter keys
+var (
+	keyInflationRateChange = KeyInflationRateChange()
+	keyInflationMax        = KeyInflationMax()
+	keyInflationMin        = KeyInflationMin()
+	keyGoalBonded          = KeyGoalBonded()
+	keyUnbondingTime       = KeyUnbondingTime()
+	keyMaxValidators       = KeyMaxValidators()
+	keyBondDenom           = KeyBondDenom()
+)
+
+// InflationRateChange - Maximum annual change in inflation rate
+func (k Keeper) InflationRateChange(ctx sdk.Context) (res sdk.Rat) {
+	k.paramstore.Get(ctx, keyInflationRateChange, &res)
+	return
+}
+
+// InflationMax - Maximum inflation rate
+func (k Keeper) InflationMax(ctx sdk.Context) (res sdk.Rat) {
+	k.paramstore.Get(ctx, keyInflationMax, &res)
+	return
+}
+
+// InflationMin - Minimum inflation rate
+func (k Keeper) InflationMin(ctx sdk.Context) (res sdk.Rat) {
+	k.paramstore.Get(ctx, keyInflationMin, &res)
+	return
+}
+
+// GoalBonded - Goal of percent bonded atoms
+func (k Keeper) GoalBonded(ctx sdk.Context) (res sdk.Rat) {
+	k.paramstore.Get(ctx, keyGoalBonded, &res)
+	return
+}
+
+// UnbondingTime
+func (k Keeper) UnbondingTime(ctx sdk.Context) (res int64) {
+	k.paramstore.Get(ctx, keyUnbondingTime, &res)
+	return
+}
+
+// MaxValidators - Maximum number of validators
+func (k Keeper) MaxValidators(ctx sdk.Context) (res uint16) {
+	k.paramstore.Get(ctx, keyMaxValidators, &res)
+	return
+}
+
+// BondDenom - Bondable coin denomination
+func (k Keeper) BondDenom(ctx sdk.Context) (res string) {
+	k.paramstore.Get(ctx, keyBondDenom, &res)
+	return
+}
+
+// Get all parameteras as types.Params
+func (k Keeper) GetParams(ctx sdk.Context) (res types.Params) {
+	res.InflationRateChange = k.InflationRateChange(ctx)
+	res.InflationMax = k.InflationMax(ctx)
+	res.InflationMin = k.InflationMin(ctx)
+	res.GoalBonded = k.GoalBonded(ctx)
+	res.UnbondingTime = k.UnbondingTime(ctx)
+	res.MaxValidators = k.MaxValidators(ctx)
+	res.BondDenom = k.BondDenom(ctx)
+	return
+}
+
+// Need a distinct function because setParams depends on an existing previous
+// record of params to exist (to check if maxValidators has changed) - and we
+// panic on retrieval if it doesn't exist - hence if we use setParams for the very
+// first params set it will panic.
+func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
```

Comment:
Can we make it clear that it's for testing & is inefficient?  maybe SetParamsSlow(...)

Reply:
SetNewParams() below will call paramstore.Set() on each types.Param field... which is probably not what we want except for testing or upon genesis...  And currently SetParams calls SetNewParams, so this isn't as optimal...

Maybe this should be renamed to "SetParamsForTesting", and in production code prefer individual Set methods?  Then, we can have a "DidUpdateParam(ctx, key, oldValue, newValue)", and we can move the conditional below (`if k.MaxValidator(ctx)... { ... }`) in there, under a giant switch statement that switches on key.

Reply:
We should create a separate issue for it instead of delaying this PR.
TODO create PR to discuss this issue.

Reply:
If this isn't obvious, then we should create a new issue, because this PR still works.

Reply:
It was originally came from https://github.com/cosmos/cosmos-sdk/blob/develop/x/stake/keeper/keeper.go#L58 so I'm not sure is it safe to modify. Opening an issue for it looks good for me.